### PR TITLE
Allow reordering of custom entry types fields

### LINKED ISSUE #6338: Dialog "Customize Entry Types" - drag&drop not working - adding new field buggy
Luckily the "Customize Entry Types"-Dialog ist back! Unfortunately it is not (yet) working thoroughly:
* the "plus"-button for fields is grey and can therefore not be pressed, although I have entered a fieldname in the dropdown-box - this only changes after one clicks into the "new entry type"-box at least once
* the drag and drop to change the order isn't working at all (or at least I cannot figure out where to touch the line to drag&drop)

I have experienced this using:
JabRef 5.0--2020-03-06--2e6f433
Windows 10 10.0 amd64 
Java 13.0.2

I have tested the latest development version and the problem persists:
JabRef 5.1--2020-04-23--bbea7df
Windows 10 10.0 amd64 
Java 14.0.1

### PR DESCRIPTION
**Pull request by Siedlerchr**
Additionally, restore the old Multiline Property and make it configurable.

Current status:
- Adding/Removing/Reordering works
- Removing standard entry types has no effect (they are always readded if deleted

- Sort fields alphabetically? Would that effect the order? Needs investigation
==> cannot sort. it influences the save order 

Refs #5230 and #4373

Fixes  #6338

- [ ] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for bigger UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not: Issue created at <https://github.com/JabRef/user-documentation/issues>.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
LinkedHashSet is no solution either, as the standard fields are always inserted first.
TODO check it again

**Comment by Siedlerchr** [comment]
@tobiasdiez I currently have the problems that in my HashMap which contains entry types with their corresponding fields, the fields order is not updated if I change them via drag and drop in the main table.
This is logical, because the fieldsWithTypes  is not an Observable.
Any idea how I can do this with some Binding Foo? Can I add some ObservabeList to that map?

https://github.com/JabRef/jabref/pull/6152/files#diff-8313b29670cb7c9bc50c020e1b2b5f17L80-R101

**Comment by tobiasdiez** [comment]
I'm not sure what exactly you mean. A `Map<EntryType, ObservableList<Field>>` does not work?

**Comment by Siedlerchr** [comment]
I tired it, but its a bit more complicated I think (Or maybe I am thinking too complicated)

I would need to bind a ListProperty to the values Collection of my map
and then I would need to change that ListProperty (fieldsForType) content based on my selected entry types. Currently I am doing this more or less manually.


```
    private final ListProperty<FieldViewModel> fieldsForType;

    private final ObservableList<FieldViewModel> allFieldsForType = FXCollections.observableArrayList(extractor -> new Observable[] {extractor.fieldName(), extractor.fieldType()});

      for (BibEntryType entryType : allTypes) {
            List<FieldViewModel> fields = entryType.getAllFields().stream().map(bibField -> new FieldViewModel(bibField.getField(), entryType.isRequired(bibField.getField()), bibField.getPriority(), entryType)).collect(Collectors.toList());
            typesWithFields.put(entryType, FXCollections.observableArrayList(fields));
        }
        

        this.fieldsForType = new SimpleListProperty<>(allFieldsForType);

        EasyBind.subscribe(selectedEntryTypes, type -> {
            if (type != null) {
                allFieldsForType.setAll(typesWithFields.get(type));
            }
        });
```

**Comment by tobiasdiez** [comment]
I guess the following might work:
- Create your `Map<EntryType, ObservableList<Fields>> entryTypesWithFields`. 
- When the user selects an `entryType` then set `fieldsForType` to `entryTypesWithFields.get(entryType)`. It is important that you set really the observable list and not the contents of the list (i.e. not `setAll`).

By the way, the whole map construction seems to be overly complicated. Why not create a `EntryTypeViewModel` that wraps a given entry type and also contains an observable list of fields. Then you can have a simple observable list of EntryTypeViewModels, which you can bind to the list and then bind the contents of the field list to the observable list of the current entry type.

File: src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java [hunk 1]
```diff
@@ -141,6 +141,7 @@ private void setupTable() {
         //Here we would need to select the
         EasyBind.subscribe(viewModel.selectedEntryTypeProperty(), type -> {
             if (type != null) {
+                fields.itemsProperty().unbindBidirectional(type.fields());
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java:144]: @tobiasdiez I followed your approach, however, sometimes the field list  on the right is not updated.
I tried unbinding/binding again but doesn't help. Any idea  how to fix this?

File: src/main/java/org/jabref/gui/DragAndDropDataFormats.java [hunk 2]
```diff
@@ -11,6 +11,7 @@
  */
 public class DragAndDropDataFormats {
 
+    public static final DataFormat FIELD = new DataFormat("dnd/org.jabref.model.entry.field.Field");
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/DragAndDropDataFormats.java:14]: For this JabRef-intern drag&drops, the https://github.com/JabRef/jabref/blob/master/src/main/java/org/jabref/gui/util/CustomLocalDragboard.java is easier to use.

File: src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeDialogViewModel.java [hunk 3]
```diff
@@ -55,19 +51,17 @@ public Field fromString(String string) {
...
     private final ObjectProperty<Field> selectedFieldToAdd = new SimpleObjectProperty<>();
     private final StringProperty entryTypeToAdd = new SimpleStringProperty("");
     private final ObservableList<BibEntryType> allEntryTypes;
-    private final ObservableList<FieldViewModel> allFieldsForType = FXCollections.observableArrayList(extractor -> new Observable[] {extractor.fieldName(), extractor.fieldType()});
     private final ObjectProperty<Field> newFieldToAdd = new SimpleObjectProperty<>();
     private final BibDatabaseMode mode;
-    private final Map<BibEntryType, List<FieldViewModel>> typesWithFields = new HashMap<>();
     private final List<BibEntryType> typesToRemove = new ArrayList<>();
 
+    private final ListProperty<CustomEntryTypeViewModel> entryTypesWithFields = new SimpleListProperty<>(FXCollections.observableArrayList());
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeDialogViewModel.java:63]: Try to make this a simple ObservableList.

File: src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeDialogViewModel.java [hunk 4]
```diff
@@ -55,19 +51,17 @@ public Field fromString(String string) {
         }
     };
 
-    private final ListProperty<BibEntryType> entryTypes;
-    private final ListProperty<Field> fields;
-    private final ObjectProperty<BibEntryType> selectedEntryTypes = new SimpleObjectProperty<>();
-    private final ListProperty<FieldViewModel> fieldsForType;
+    private final ListProperty<Field> fieldsForAdding;
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeDialogViewModel.java:54]: Here too: simply use an ObservableList

File: src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeViewModel.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import javafx.beans.property.SimpleObjectProperty;
+import javafx.collections.FXCollections;
+import javafx.collections.ListChangeListener;
+
+import org.jabref.model.entry.BibEntryType;
+
+public class CustomEntryTypeViewModel {
+
+    private final ObjectProperty<BibEntryType> entryType = new SimpleObjectProperty<>();
+    private final ListProperty<FieldViewModel> fields;
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeViewModel.java:19]: again, ObservableList should suffice

File: src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeViewModel.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    private final List<FieldViewModel> fieldsToRemove = new ArrayList<>();
+
+    public CustomEntryTypeViewModel(BibEntryType entryType) {
+        this.entryType.set(entryType);
+
+        List<FieldViewModel> allFieldsForType = entryType.getAllFields().stream().map(bibField -> new FieldViewModel(bibField.getField(), entryType.isRequired(bibField.getField()), bibField.getPriority(), entryType)).collect(Collectors.toList());
+
+        fields = new SimpleListProperty<>(FXCollections.observableArrayList(allFieldsForType));
+
+        fields.addListener((ListChangeListener<? super FieldViewModel>) change -> {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeViewModel.java:29]: This looks a bit complicated. Why not simply compare/overwrite on-save the old list of fields with the new one?

File: src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java (hunk 1 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java:144]: This should be `fields.setItems(type.fields())` after converting all the above list properties to observable lists.

Moreover, I would put this subscription thing in the `viewModel` and expose a new list `viewModel.fieldsForSelectedType`.

**Comment by Siedlerchr** [comment]
@tobiasdiez  Sometimes when I go up and down and switch between the entry types, suddenly the fields are empty. Any idea what this means or why this occurs?

File: src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeDialogViewModel.java [hunk 7]
```diff
@@ -55,18 +52,14 @@ public Field fromString(String string) {
...
+    private final ObjectProperty<CustomEntryTypeViewModel> selectedEntryTypes = new SimpleObjectProperty<>();
     private final ObjectProperty<Field> selectedFieldToAdd = new SimpleObjectProperty<>();
     private final StringProperty entryTypeToAdd = new SimpleStringProperty("");
     private final ObservableList<BibEntryType> allEntryTypes;
-    private final ObservableList<FieldViewModel> allFieldsForType = FXCollections.observableArrayList(extractor -> new Observable[] {extractor.fieldName(), extractor.fieldType()});
     private final ObjectProperty<Field> newFieldToAdd = new SimpleObjectProperty<>();
     private final BibDatabaseMode mode;
-    private final Map<BibEntryType, List<FieldViewModel>> typesWithFields = new HashMap<>();
-    private final List<BibEntryType> typesToRemove = new ArrayList<>();
+    private final ListProperty<CustomEntryTypeViewModel> entryTypesWithFields = new SimpleListProperty<>(FXCollections.observableArrayList(extractor -> new Observable[] {extractor.entryType(), extractor.fields()}));
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeDialogViewModel.java:62]: observable list should be sufficient here.

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeDialogViewModel.java:62]: Nope, it's used in a binding and bind() only accepts a ListProperty

Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeDialogViewModel.java:62]: `entryTypes.setItems(viewModel.entryTypes());` should work if `entryTypes` is an observable list. No need for the bind + property.

**Comment by Siedlerchr** [comment]
The issue with the delete on change is fixed.
Now I think I encountered a bug with the EntryTypesManager parsing. 
If a custom entry type has only required or optional fields, (e.g. one of the fields array is empty), a "ghost" bib field is created.
Need to invesitgate this 

grafik

File: src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java [hunk 8]
```diff
@@ -128,16 +149,88 @@ private void setupTable() {
                 .install(fieldTypeActionColumn);
 
         viewModel.newFieldToAddProperty().bindBidirectional(addNewField.valueProperty());
-        fields.itemsProperty().bindBidirectional(viewModel.fieldsforTypesProperty());
+
+        // Here we would need to select the fields
```
Comment (koppor) [inline src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java:153]: I don't get this comment. --> I read it as "We do not need to select the fields and do something different, because of some idea" 😇 

It seems that this is a necessary comment. Could you maybe rephrase it?

File: src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java [hunk 9]
```diff
@@ -128,16 +149,88 @@ private void setupTable() {
...
+}
+
+    private void handleOnDragOver(FieldViewModel originalItem, DragEvent event) {
+        if ((event.getGestureSource() != originalItem) && event.getDragboard().hasContent(DragAndDropDataFormats.FIELD)) {
+            event.acceptTransferModes(TransferMode.MOVE);
+        }
+    }
+
+    private void handleOnDragDetected(TableRow<FieldViewModel> row, FieldViewModel fieldViewModel, MouseEvent event) {
+        // Start drag'n'drop
```
Comment (koppor) [inline src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java:175]: Do we really need that comment? - Where is "stop drag'n'drop"?

File: src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java [hunk 10]
```diff
@@ -128,16 +149,88 @@ private void setupTable() {
...
+        event.consume();
+    }
+
+    private void handleOnDragDropped(TableRow<FieldViewModel> row, FieldViewModel originalItem, DragEvent event) {
+        boolean success = false;
+
+        ObservableList<FieldViewModel> items = fields.itemsProperty().get();
+
+        if (localDragboard.hasType(FieldViewModel.class)) {
+
```
Comment (koppor) [inline src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java:194]: No empty line here.

File: src/test/java/org/jabref/logic/bibtex/BibEntryWriterTest.java [hunk 11]
```diff
@@ -272,8 +273,8 @@ void roundTripWithCamelCasingInTheOriginalEntryAndResultInLowerCase() throws IOE
         String bibtexEntry = OS.NEWLINE + "@Article{test," + OS.NEWLINE +
                 "  Author                   = {Foo Bar}," + OS.NEWLINE +
                 "  Journal                  = {International Journal of Something}," + OS.NEWLINE +
-                "  Note                     = {some note}," + OS.NEWLINE +
```
Comment (koppor) [inline src/test/java/org/jabref/logic/bibtex/BibEntryWriterTest.java:275]: This change is OK for me. `Number` really belongs to a journal. More than `Note`

File: src/test/resources/testbib/bibWithUserComments.bib [hunk 12]
```diff
@@ -33,12 +33,12 @@ @Comment this
 @INPROCEEDINGS{1137631,
   author = {Gustav Bostr\"{o}m and Jaana W\"{a}yrynen and Marine Bod\'{e}n and
 	Konstantin Beznosov and Philippe Kruchten},
-  title = {Extending XP practices to support security requirements engineering},
```
Comment (koppor) [inline src/test/resources/testbib/bibWithUserComments.bib:36]: Why is here the `title` **after** `booktitle` and in `bibWithUserCommentAndEntryChange.bib`, the `title` **before** `booktitle` (which I like)

We had huge debates about field ordering. I think, we agreed that `author` comes first and then `title`. Then the other metadata.

File: src/test/resources/testbib/bibWithUserComments.bib [hunk 13]
```diff
@@ -33,12 +33,12 @@ @Comment this
...
   booktitle = {SESS '06: Proceedings of the 2006 international workshop on Software
 	engineering for secure systems},
+  title = {Extending XP practices to support security requirements engineering},
   year = {2006},
-  pages = {11--18},
```
Comment (koppor) [inline src/test/resources/testbib/bibWithUserComments.bib:40]: This is also inconsitent to the other file.

Is this file broken maybe? - Can the change be reverted?

File: src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java (hunk 8 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/customentrytypes/CustomizeEntryTypeDialogView.java]: It was an obsolete comment I forgot to rmeove

**Comment by Siedlerchr** [comment]
TODO: 
1. Modify type article, add field langid
2. Make optional 
3. Move after number
4. Save
5. Reopen dialog
=> langid field is no longer after number, it has moved to a different position

Problem possible: in BibEntryTypesManager;
```
   public SortedSet<BibEntryType> getAllTypes() {
            SortedSet<BibEntryType> allTypes = new TreeSet<>(customOrModifiedType);
            allTypes.addAll(standardTypes);
            return allTypes;
        }
```

**Comment by Siedlerchr** [comment]
Yeah! I found the bug:
grafik
